### PR TITLE
Experiment building blocks: Base + Core Mixins

### PR DESCRIPTION
**Pull request by mrcslws**
This change adds a BaseExperiment that is primarily an interface class with very little logic, and it moves reusable logic into mixins and "core mixins". A core mixin is a new tier of mixin that is allowed to add additional APIs to the class.

This initial change factors Distributed and StepBasedLogging logic into core mixins, and ExtraValidations logic into a regular mixin. More reusable logic will be refactored into mixins (regular and core) in future changes.

This is building up to a future change that makes the CL experiments stop inheriting from SupervisedExperiment. When this inheritance link is removed, the CL experiments will become simpler (e.g. they won't need explicit aggregate_results methods)

### CODE REVIEW

**Comment by mrcslws** [comment]
Based on feedback, I changed the name from "expansion" to "core mixin". I amended the commit.

**Comment by mrcslws** [comment]
FYI I am prototyping a change that uses interface classes with abstract methods to properly document dependencies, rather than using this duck-typing approach. You can hold off on reviewing this change.

File: nupic/research/frameworks/vernon/mixins/extra_validations.py
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/mixins/extra_validations.py:84]: Could you add a docstring to this? Specifically, it could be helpful to help distinguish it from `StepBasedLogging. expand_result_to_time_series`.

File: nupic/research/frameworks/vernon/core_mixins/step_based_logging.py
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/core_mixins/step_based_logging.py:130]: I see `k_mapping` is defined here and in `ExtraValidation`. Maybe we can make this into a class function (which I believe we've discussed before)?

It would help reduce redundancy, for one. But, more importantly, it leaves `StepBasedLogging` more agnostic to what's being logged at every step. For instance, the `ContinualLearningExperiment` also trains in batches but uses different metrics.

File: nupic/research/frameworks/vernon/mixins/extra_validations.py
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/mixins/extra_validations.py:104]: You include `if "aggregate_results" in eo` to pose the parent functionality as optional. However, here you assume the parent has the functionality. 

The fix is easy: `if hasattr(super(), aggregate_results)` 
But it's weird to think about its implications. This, in a sense, is optional inheritance. I'm not sure if this a red flag.

Maybe we can throw a warning when `ExtraValidations.aggregate_results` is called yet `super().aggregate_results` doesn't exist.

File: nupic/research/frameworks/vernon/experiments/base_experiment.py
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/experiments/base_experiment.py:177]: I think `create_optimizer` is here unintentionally.

File: nupic/research/frameworks/vernon/experiments/base_experiment.py
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/experiments/base_experiment.py:181]: Are you missing `should_stop` in this list? Is this intentional?

File: nupic/research/frameworks/vernon/experiments/base_experiment.py
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/experiments/base_experiment.py:128]: Generally, the pre_experiment could be anything. For instance, one could record the sparsity level of the model at every epoch.

File: nupic/research/frameworks/vernon/experiments/base_experiment.py
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/experiments/base_experiment.py:182]: Are you missing `insert_pre_experiment_result` in this list? Is that also intentional?

File: nupic/research/frameworks/vernon/experiments/base_experiment.py
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/experiments/base_experiment.py:172]: Generally, could you help me understand when something deserves to go into the execution order? Maybe we should describe it in the docstring here.

**Review by mvacaporale — CHANGES_REQUESTED** [review]
Looks good. Thanks for all the refactoring and tidying up.

High-level comment as we move forward: My one worry (and @subutai has expressed this to me as well) is whether the relationships between the mixins are easily understood and, more importantly, implemented in a way to minimize the risk of errors. 

Of course, there are checks in this PR, such as checking for `current_timestep` when needing `StepBasedLogging,` but it's worth reflecting on how these mixins will interact with each other as well as their intended interactions.

It's hard to say much more than that for now. Moving forward, it may be worth it to create tests for some canonical/likely mixtures of mixins to ensure what things we hope will work well together, in fact, do so.

Update: I wrote this after reading your last comment. Your "interface classes with abstract methods idea" may work well for the concern expressed above.

**Comment by mvacaporale** [comment]
> FYI I am prototyping a change that uses interface classes with abstract methods to properly document dependencies, rather than using this duck-typing approach. You can hold off on reviewing this change.

I just read this as I finished up the review. That may affect my comments. Please interpret them as you see fit given your changes.

File: nupic/research/frameworks/vernon/core_mixins/distributed.py
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/core_mixins/distributed.py:71]: Just a side comment, not a change request: there is a small but so far neglectable chance this port will be taken. Might become an issue if extending framework to more users to throw an error if that is the case or choose a free port instead.

File: nupic/research/frameworks/vernon/core_mixins/step_based_logging.py
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/core_mixins/step_based_logging.py:30]: This looks great. In the future I was wondering if we could refactor in such a way that user can define the level he wants to log, eg: step, epoch, task, bundle of tasks (as in meta learning). I haven't given a lot of thought about this, and I don't know a good approach to do this given our current framework. But maybe we can pingpong ideas some day.

File: nupic/research/frameworks/vernon/experiments/base_experiment.py
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/experiments/base_experiment.py:157]: A maybe controversial opinion: should we create an option for the experiment to train itself, without the Trainable? That would be a very vanilla function (`while not should stop: run epoch; print results`). I think that would make a lot easier for beginners to run simple experiments on notebooks, just instantiate an experiment class and call `exp.train()`.

No changes to the whole Trainable thing, it is just becomes another option. Lots of frameworks have more than one approach of doing something, usually a simpler version and a more complex one for power users (Keras for example).

File: nupic/research/frameworks/vernon/mixins/extra_validations.py
Comment (lucasosouza) [reply nupic/research/frameworks/vernon/mixins/extra_validations.py:84]: I was a bit confused by the term `time series`, at least for me it relates for a method to handle datasets with a temporal dimension (like in RNNs / HTMs).

File: nupic/research/frameworks/vernon/mixins/extra_validations.py
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/mixins/extra_validations.py:31]: I would suggest a more specific name like `ExtraValidationsPerEpoch` or something similar. `ExtraValidations` is quite generic, and I could only get its meaning by going through the docstring and code. By the name only I initially thought it was a way to add more performance measures to the validation process.

File: nupic/research/frameworks/vernon/trainables.py
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/trainables.py:331]: Can't we move this to the Experiment class instead, and ditch this Trainable as well? This can be an attribute or a function, I think an attribute makes more sense.

**Review by lucasosouza — APPROVED** [review]
This looks great, thanks. Left some minor comments.

File: nupic/research/frameworks/vernon/trainables.py
Comment (mvacaporale) [reply nupic/research/frameworks/vernon/trainables.py:331]: I added that, although I'm not sure it's needed. Isn't the stop condition adequately covered by `should_stop`?

File: nupic/research/frameworks/vernon/mixins/extra_validations.py
Comment (mrcslws) [reply nupic/research/frameworks/vernon/mixins/extra_validations.py]: I added some code comments. (At the moment, I'm not quite sure how I should think about docstrings on an overriden method. Docstrings on interfaces are surfaced to IDEs, etc., so I'm making sure the interfaces have good docstrings.)

I'll think about new terminology in the future... the "time series" terminology has been in this code for many months. (And this change is already big.)

File: nupic/research/frameworks/vernon/mixins/extra_validations.py
Comment (mrcslws) [reply nupic/research/frameworks/vernon/mixins/extra_validations.py]: Yeah, these "optional" methods were gross. I ended up doing a big refactor to get rid of them (creating `vernon.distributed` to capture all the distributed logic in subclass mixins)

File: nupic/research/frameworks/vernon/experiments/base_experiment.py
Comment (mrcslws) [reply nupic/research/frameworks/vernon/experiments/base_experiment.py]: Good point, I added it to the interface docstring.

File: nupic/research/frameworks/vernon/trainables.py
Comment (mrcslws) [reply nupic/research/frameworks/vernon/trainables.py]: Ok, I got rid of this. Yeah, it's covered by `should_stop`. If we want to make future changes to the stopping logic, we can. This change was more of a refactor, I didn't put much thought into what is the best way to specify stopping criteria.

**Comment by mrcslws** [comment]
I made a pretty significant-sized change. The "Distributed" core mixin, the "experiment_base.py", and the "evaluation_metrics.py" are now in a `experiments/components` subfolder. It might be best to think of them as experiment components that are combined via multiple inheritance (though you can also think of them as mixins if that's useful). `step_based_logging.py` is now a mixin, and the other mixins that rely on it inherit from it.

As discussed, I added the notion of "interfaces".

Additionally, I split out the "distributed" experiments and mixins into a separate `vernon.distributed` module.

```python
# Option 1
from vernon import experiments, mixins  
# Option 2
from vernon.distributed import experiments, mixins  
```

It's now easy to write Vernon experiments without needing to make them distributed. I refactored the Supervised, CL, and Meta CL experiments into regular and distributed versions, and I refactored the trainables to support either case. With this refactor, the distributed Meta CL experiment is a bit hacky, it might warrant some future cleanup.

All these changes still fit with the original goal of this change: factoring out `step_based_logging` and `distributed` logic so that experiments don't have to wrestle with them unless they want to.

File: nupic/research/frameworks/vernon/mixins/step_based_logging.py
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/mixins/step_based_logging.py:115]: Can we add a TODO or even a mental note for later to be able to define this `k_mapping` via the config or a classmethod?

**Comment by mvacaporale** [comment]
Still reviewing. Looks good so far, but I'll have to finish tomorrow.

File: nupic/research/frameworks/vernon/trainables.py
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/trainables.py:278]: Could you add a brief docstring to this stating it's used to extend the experiment classes? Or maybe it's obvious enough. Up to you.

File: nupic/research/frameworks/vernon/trainables.py
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/trainables.py:171]: Is there an easy way to use this trainable with a `DataParallel` model? If anything, it would be useful for getting the benefit of multiple GPUs without the hassle of aggregating results. We can save this for later though when if the need ever arises (Our CL and MCL experiments use just one GPU at the moment, for instance.)

**Review by mvacaporale — APPROVED** [review]
All in all looks good. I only left minor comments.

If anything, I do worry about the learning curve. I'm really curious to see your final PR. Hopefully, we can have an easy to use and minimal experiment class. At the very least, I'm sure we can also make one and a simple how-to-run example, as Lucas has suggested in the past.

File: nupic/research/frameworks/vernon/README.md
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/README.md:90]: Maybe we can add links that give an overview of interfaces and mixins. These are new concepts for me and I imagine they would be new for others even if they're familiar with OOP.

Here are some links we could consider adding:
mixins - https://realpython.com/inheritance-composition-python/#mixing-features-with-mixin-classes
interfaces - https://realpython.com/python-interface/#python-interface-overview

File: nupic/research/frameworks/vernon/trainables.py
Comment (mrcslws) [reply nupic/research/frameworks/vernon/trainables.py:171]: For now, we can do that by extending `setup_experiment` (via a subclass or a mixin). Also, I'm considering removing the DataParallel usage from the Distributed* classes -- that doesn't seem like the right place for it.

**Comment by mrcslws** [comment]
@mvacaporale Thanks for the rigorous read-through, the execution orders definitely needed some cleanup. Soon I'll post a README update that gives more explanation of how Vernon uses interfaces. (You had a question about `distributed_aggregation_interface_implemented` https://github.com/numenta/nupic.research/pull/376#discussion_r514775953 so I want to make sure the README answers that question. Also, it will explain why experiment_base.py is a component -- a Vernon experiment doesn't necessarily need to use it, it just needs to implement the Experiment interface _somehow_)

File: nupic/research/frameworks/vernon/README.md
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/README.md:92]: Maybe we should make it clear these are not just tricks but standards enforced in this library.

File: nupic/research/frameworks/vernon/distributed/experiments/cl_experiment.py
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/distributed/experiments/cl_experiment.py:40]: AFAIK we don't we have any tests for ContinualLearning experiment yet, neither unit or regression. That is not the subject of this PR, but just a note that while we don't have tests we can unintentionally break some code with something as simple as a typo in a refactoring. 

I can test it later after you merge to make sure that is not the case (and probably see if I write some unit tests in the process).

File: nupic/research/frameworks/vernon/distributed/experiments/components/distributed_base.py
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/distributed/experiments/components/distributed_base.py:33]: You explained to us the distinction between Implements and Requires "tags" and why do we use them, but maybe we can add a line about it in the documentation as well, if we haven't already (that first README can be a good place). 

It might sound obvious after some thinking, but I don't think it is at first glance to someone who is skimming through the library to learn how it works.

File: nupic/research/frameworks/vernon/distributed/experiments/supervised_experiment.py
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/distributed/experiments/supervised_experiment.py:54]: As Subutai pointed out, this might not be the most efficient approach for a single machine, although we don't know that for sure without testing. Not relevant for Imagenet but most of the experiments we have been running in Continual and MetaContinual Learning are in single instances.

It does make it more convenient but something we can keep in mind in the future is to test and see whether this is impacting single instance runs or not.

File: nupic/research/frameworks/vernon/interfaces/distributed_aggregation.py
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/interfaces/distributed_aggregation.py:68]: I was wondering what this method was for, but it is clear here.

File: nupic/research/frameworks/vernon/trainables.py
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/trainables.py:318]: The abbreviation Exp sounded a bit weird given in all other classes in vernon we don't refrain from using full names even if super long.

File: nupic/research/frameworks/vernon/distributed/experiments/supervised_experiment.py
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/distributed/experiments/supervised_experiment.py:47]: I was a bit confused here - the class name is SupervisedExperiment, but the docstring says Distributed SupervisedExperiment (judging from the inheritance it is the latter).

What does that mean, is Distributed the default way? How to run non-distributed Supervised Experiments?

**Review by lucasosouza — COMMENTED** [review]
Nothing that requires changes, I only have a few comments and some questions that can help me understand the changes better.

File: nupic/research/frameworks/vernon/distributed/experiments/cl_experiment.py
Comment (mrcslws) [reply nupic/research/frameworks/vernon/distributed/experiments/cl_experiment.py:40]: For what it's worth, I have been running the CL and Meta CL experiments to make sure they don't throw exceptions. But I haven't been verifying correct results.

File: nupic/research/frameworks/vernon/distributed/experiments/components/distributed_base.py
Comment (mrcslws) [reply nupic/research/frameworks/vernon/distributed/experiments/components/distributed_base.py:33]: Definitely agree, I was in the process of writing that.

File: nupic/research/frameworks/vernon/distributed/experiments/supervised_experiment.py
Comment (mrcslws) [reply nupic/research/frameworks/vernon/distributed/experiments/supervised_experiment.py:47]: `vernon.experiments.SupervisedExperiment` is non-distributed. `vernon.distributed.experiments.SupervisedExperiment` is distributed.

Comment (lucasosouza) [reply nupic/research/frameworks/vernon/distributed/experiments/supervised_experiment.py:47]: Got it, makes sense now, thanks.

**Pull request merged**
